node/src/main.rs

```diff
@@ -44,14 +50,48 @@ use graph_server_websocket::SubscriptionServer as GraphQLSubscriptionServer;
 use graph_store_postgres::{Store as DieselStore, StoreConfig};
 
 fn main() {
+    let (shutdown_sender, shutdown_receiver) = oneshot::channel();
+    // Register guarded panic logger which ensures logs flush on shutdown
     let (panic_logger, _panic_guard) = guarded_logger();
-    register_panic_hook(panic_logger);
-    tokio::run(future::lazy(async_main))
+    register_panic_hook(panic_logger, Mutex::new(Some(shutdown_sender)));
+
+    // Create components for tokio context: multi-threaded runtime, reactor reference,
+    // executor context on the runtime, and Timer handle.
+    let runtime = tokio::runtime::Runtime::new().expect("Failed to create runtime");
+    let reactor = tokio_reactor::Handle::current();
+    let mut executor = runtime.executor();
+    let mut enter = tokio_executor::enter()
+        .expect("Failed to enter runtime executor, multiple executors at once");
+    let timer = Timer::default();
+    let timer_handle = timer.handle();
+
+    std::thread::spawn(|| {
+        shutdown_receiver
+            .wait()
+            .map(|_| {
+                let _ = runtime
+                    .shutdown_now()
+                    .wait()
+                    .expect("Failed to shutdown Tokio Runtime");
+                thread::sleep(Duration::from_millis(3000));
```

Comment:
Don't know what you observed in your tests, but the idea here is that we'd have a timeout for `shutdown_now`. If `shutdown_now` succeeds, then we should not need to call `process_exit(1)` since the process should naturally terminate with the termination of the main thread.

So we should have a timeout instead of calling `wait` directly on `shutdown_now`, and we should call `exit` only if the timeout elapses.

Reply:
I had planned on going the route you mention.  Using a `timeout` instead of `wait`, but I found that in one of my test cases (a panic in `instance_manager` around line 76) `shutdown_now` completes successfully but there is still a process running.  So instead of calling 'exit' only if it doesn't complete in a given time I allow it to complete and sleep the shutdown thread for 3 seconds before exiting.

Reply:
Revisiting the issue this week, I was not able to repeat the case of `shutdown_now()` hanging, so the timeout didn't have an effect on my test cases.

Reply:
Ah I see. So based on that experiment, having the unconditional `exit` is good. But I don't see why we need to sleep before doing that. Also it still seems good to have a timeout around `shutdown_now()` since at some point could observe it hanging.

Reply:
I added the sleep in because when testing with multiple panics (in separate threads), the sleep gives time for both to be logged before exiting.  I could go either way on this; it's not necessarily a problem that we don't see issues that occur after the original panic, but is nice to see them both.    Agreed, I'll add a timeout around `shutdown_now` (it's a little tricky using tokio_timer around for an operation that is shutting down tokio, I may create a new timer instance or use `futures-timer`.)